### ISSUE TITLE
Camera - App passed NULL surface

### ISSUE THREAD

**Issue by pedrocordeiromendes** [Redth/ZXing.Net.Mobile#550]
Hi,

I am using the latest version of ZXing.Net.Mobile (v2.2.9) in my Android application to perform QR code scanning.

Even though I am being able to scan every QR code successfully so far, the app freezes randomly after the scan and the device logs a camera related error - "app passed NULL surface". After this message the app stays frozen for a few seconds until another message is logged into the device logger - "[W/CameraBase] Camera service died!". After this happens the app resumes its normal behavior.

How can I solve this problem?

In my core project I have an event to initiate the scan:

```
private void StartScan()
        {
            try
            {
                
                MobileBarcodeScanningOptions oOptions = new MobileBarcodeScanningOptions();
                oOptions.PossibleFormats = new List<BarcodeFormat>() { BarcodeFormat.QR_CODE };
                oOptions.DelayBetweenContinuousScans = 3000;

                _oScanner = new MobileBarcodeScanner();
                _oScanner.TopText = Notices.TITLE_SCANNER_QRCODE;
                _oScanner.CancelButtonText = Notices.LABEL_CANCEL;

                _oScanner.ScanContinuously(oOptions, HandleScanResult);

            }
            catch (Exception ex)
            {
                Logger.Debug("StartScan[ScannerViewModel]", ex);
            }
        }
private void HandleScanResult(Result oResult)
        {
            try
            {
                if (oResult != null && !string.IsNullOrEmpty(oResult.Text))
                {
                   
                       //In here I am handling the result of the scan and navigating to another view
                }
            }
            catch (Exception ex)
            {
                Logger.Debug("HandleScanResult[ScannerViewModel]", ex);
            }
            finally
            {
                if (_oScanner != null)
                    _oScanner.Cancel();
            }

        }

```

Any help is appreciated.

**Comment by fredrikhaglund** [comment]
I have also seen this in the log for Android using the latest version when trying to use the Xamarin Forms wrapper ZXingScannerView and create the control at the same time as the page was created. In a similar way the MobileBarcodeScanner creates a new page/activity to show the scanner.

In my case I changed the strategy and delayed creating the scanner view until the page was visible and things became stable. So I keep my ZXingScannerView active and have no issues. Whenever I turn of continious scan I remove my control and create a new one for the next round. This appeared to help. 

Another suspect I had was what I did in my event handler. I had some async code there and this somehow caused issues. 

I would try to comment out you code that actually does the handling of the result with just some debug logging and try to see if it becomes stable. Maybe add some logging with Environment.CurrentManagedThreadId to see if you switch thread? Or are you trying to call UI code from another thread than the main thread?

Even if my scenario is very different at least this proves that you can work around this.

**Comment by fredrikhaglund** [comment]
Another thing you should try is to rollback to nuget package version [2.1.47](https://www.nuget.org/packages/ZXing.Net.Mobile/2.1.47) and see if the problem also exists there or not. 

A huge refactoring was released just a two weeks ago and new issues and some regression was introduced.

**Comment by pedrocordeiromendes** [comment]
Hi Fredrik,

Thanks for the reply.

My initial approach would fall in your scenario, since it involved calling async methods, doing work on the UI thread while waiting and then resolve a different view. But then I started to simplify the handling of the QR code but I could find this problem even when I did no handling at all. Just null reference validations and then closing the scanner.

I followed your suggestion and rolled back to version 2.1.47 and I am having better results. Not only the QR code scanning is getting much faster, but also the problem I had before seemed to vanish. 

I still notice that there is something wrong with the camera service by looking up the device logs (again this happens randomly), but the application does not freeze and keeps it's normal behavior. the error and ZXing stack trace when this sittuation occurs is:

```
05-02 18:30:56.825 D/ZXing.Net.Mobile( 5892): Found 0 cameras...
05-02 18:30:56.825 D/ZXing.Net.Mobile( 5892): Finding Back camera failed, opening camera 0...
05-02 18:30:56.835 W/CameraBase( 5892): An error occurred while connecting to camera: 0
05-02 18:30:56.845 W/CameraBase( 5892): Camera service died!
05-02 18:30:56.845 E/ZXING   ( 5892): Setup Error: Java.Lang.RuntimeException: Fail to connect to camera service
05-02 18:30:56.845 E/ZXING   ( 5892):   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Users/builder/data/lanes/4009/3a62f1ea/source/mono/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:143 
05-02 18:30:56.845 E/ZXING   ( 5892):   at Java.Interop.JniEnvironment+StaticMethods.CallStaticObjectMethod (Java.Interop.JniObjectReference type, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x00082] in /Users/builder/data/lanes/4009/3a62f1ea/source/Java.Interop/src/Java.Interop/Java.Interop/JniEnvironment.g.cs:12649 
05-02 18:30:56.845 E/ZXING   ( 5892):   at Java.Interop.JniPeerMembers+JniStaticMethods.InvokeObjectMethod (System.String encodedMember, Java.Interop.JniArgumentValue* parameters) [0x0001b] in /Users/builder/data/lanes/4009/3a62f1ea/source/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniStaticMethods.cs:97 
05-02 18:30:56.845 E/ZXING   ( 5892):   at Android.Hardware.Camera.Open (System.Int32 cameraId) [0x00017] in /Users/builder/data/lanes/4009/3a62f1ea/source/monodroid/src/Mono.Android/platforms/android-23/src/generated/Android.Hardware.Camera.cs:4478 
05-02 18:30:56.845 E/ZXING   ( 5892):   at ZXing.Mobile.ZXingSurfaceView+<SetupCamera>d__26.MoveNext () [0x0029e] in <04a6c4a6536f4e76a10998845fe4c337>:0 
05-02 18:30:56.845 E/ZXING   ( 5892):   --- End of managed Java.Lang.RuntimeException stack trace ---
05-02 18:30:56.845 E/ZXING   ( 5892): java.lang.RuntimeException: Fail to connect to camera service
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.hardware.Camera.<init>(Camera.java:568)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.hardware.Camera.open(Camera.java:405)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at md55109d95aac470e32f9a4031a908f0227.ZXingSurfaceView.n_surfaceChanged(Native Method)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at md55109d95aac470e32f9a4031a908f0227.ZXingSurfaceView.surfaceChanged(ZXingSurfaceView.java:69)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.view.SurfaceView.updateWindow(SurfaceView.java:740)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.view.SurfaceView$3.onPreDraw(SurfaceView.java:215)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:1014)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2599)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1524)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7520)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:911)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.view.Choreographer.doCallbacks(Choreographer.java:686)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.view.Choreographer.doFrame(Choreographer.java:622)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:897)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.os.Handler.handleCallback(Handler.java:739)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.os.Handler.dispatchMessage(Handler.java:95)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.os.Looper.loop(Looper.java:148)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at android.app.ActivityThread.main(ActivityThread.java:7303)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at java.lang.reflect.Method.invoke(Native Method)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
05-02 18:30:56.845 E/ZXING   ( 5892): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
05-02 18:30:56.845 D/ZXing.Net.Mobile( 5892): Checking android.permission.CAMERA...
```
After this happens, the camera takes a little longer to start since the camera service gets restarted.
Is there anything I should be doing to prevent this error?

**Comment by BladeMF** [comment]
I too have this error, but after IsScanning=false, IsScanning=true switch. The image freezes and that's that. Downgrading solved it.

**Comment by nishantamberkar** [comment]
Hi frnds,
I am using zxing scanner view in my project (xamarin forms). It working fine till android version 7.0 but higher version like 7.1.1 and above my app crashed.
after scanning i got scan result but after 5-6 sec app got crashed.Any solution ?
Exception is Camera : App passed null surface.
Any help is appreciated.
thanks

**Comment by philmoz** [comment]
Deleting line 137 (Camera.SetPreviewDisplay(null);) from ZXing.Net.Mobile.Android/CameraAccess/CameraController.cs seems to fix this issue for me.

According to the Android documentation the parameter to SetPreviewDisplay must be a fully initialised SurfaceHolder object, so I don't think passing null is valid.

**Comment by mazloumi** [comment]
HI, I am also facing it seems a similar issue. My app uses ZXing.Net.Mobile.Forms version 2.3.2. I tried to downgrade but without success. The ZXingScannerView scans the code successfully and calls OnScanResult. In the debug console it says

[ZXing.Net.Mobile] Barcode Found: 990616200
[Camera] app passed NULL surface

and then I get a NullReferenceException and the app freezes.

Any suggestions?

**Comment by mazloumi** [comment]
In the OnScanResult handler I replace the ZXingScannerView, which resolves the issue

**Comment by dberroa** [comment]
@philmoz Does deleting that line cause any other effects? Like would something else not work because of that deletion? I'm wondering why it was there if it causes this error.

I have yet to pull the project down to try deleting that myself to see if it works.

**Comment by philmoz** [comment]
So far I have not had any crashes since removing the Camera.SetPreviewDisplay(null); line.

I suspect it was added to 'clean up' the camera object when the scanning stops.

**Comment by dberroa** [comment]
I'm not to familiar with github so I may be saying this wrong, but do you think if it works that you should do a pull (push?) request to get it merged?

**Comment by bobvaselaar** [comment]
I'm having this issue as well. Any word on the status of fixing this? Thanks!

**Comment by yang2312** [comment]
I'm using version 2.4.1 and still facing this problem..could you guys please update the nuget version for this issue? Thanks

**Comment by BStrat77** [comment]
This is causing some issues for me as well. A publish of the NuGet would really help us out.

**Comment by dberroa** [comment]
+1 for publish.

**Comment by vjsystematix** [comment]
I too had the same problem, Scan works but when i try to capture the image I got same error as mentioned above.

Any help is really appereciated!

**Comment by BStrat77** [comment]
The only way I have been able to get around it is to Unload and Load the controls from a Grid each time I set IsScanning to False. I keep handles to the controls with member variables and manually hook up the event. It took me a while to get it working correctly, you might have to fill in a few parts of the code but this should help. It's slow to reload depending on the phone, but at least it's working for now until the new package is built.
```
private void UnloadCameraScanner()
{
    if (_cameraScanner != null)
    {
        ScannerGrid.Children.Remove(_cameraScannerOverlay);
        ScannerGrid.Children.Remove(_cameraScanner);
        _cameraScannerOverlay = null;
        _cameraScanner = null;
    }
}

private void LoadCameraScanner()
{
    UnloadCameraScanner();

    _cameraScanner = new ZXingScannerView
    {
        Options = _vm.ScanOptions,
    };
    _cameraScanner.SetBinding(ZXingScannerView.IsAnalyzingProperty, "IsAnalyzing");
    _cameraScanner.SetBinding(ZXingScannerView.IsScanningProperty, "IsScanning");

    _cameraScanner.OnScanResult += (result) => { HandleScanResult(result.Text); };
    ScannerGrid.Children.Add(_cameraScanner, 0, 0);

    _cameraScannerOverlay = new ZXing.Net.Mobile.Forms.ZXingDefaultOverlay
    {
        TopText = "Top Text",
        BottomText = "Bottom Text",
        ShowFlashButton = false
    };

    ScannerGrid.Children.Add(_cameraScannerOverlay, 0, 0);
}

<Grid x:Name="ScannerGrid" Grid.Row="0" VerticalOptions="FillAndExpand" HorizontalOptions="FillAndExpand">
   
</Grid>
```

**Comment by Redth** [comment]
Thanks for reporting this issue!  Unforunately it took me way too long to respond 😭.  Sorry, I apologize!  Recently the source code for this project was completely refactored to modernize it.  Many PR's were included in this effort, and many bugs were hopefully fixed.  Please try out the latest 3.x series of NuGet packages (currently in prerelease).  To try and make the project more maintainable in my spare time going forward, I've decided to close all existing issues to start with a clean slate.  If you're still experiencing this issue on the newest version, please open a new issue with as much detail as possible.  Thank you for your patience and understanding!  Happy scanning!

**Comment by udayaugustin** [comment]
I face the same issue in my Xamarin forms project. I have been using the Zxing scanner view for the scanner control.

In my case the issue I have pushed the scanner page inside the Navigation page. 
Once the scanning is done, I have set the **IsScanning = false**.  Then I have popped the page.

Due to this, the Null surface is passed to the camera.
This caused the null reference exception.

**To Fix this:** 
I have removed the **IsScanning = false** statement. Then it worked perfectly.

**Note:** 
I didn't face this issue when I used the Modal page for the scanner. This happened only when I pushed the normal page in the navigation stack.
